WIP: MOxUnit support for tests

nno opened the pull request:
This PR is Work In Progress (WIP) and comments at this stage are appreciated.

- It introduces a new option ``ft_test moxunit_run``and provides similar functionality as ``ft_test run``, but by extending the MOxUnit testing framework. Supported functionality includes filtering on walltime, memory usage, presence of dccnpath, and other dependencies; and uploading to the fieldtrip dashboard.
- Tests can run on Octave, and currently there is support for two online, free and open source continuous integration (CI) services, travis and shippable. Example output:
  * https://travis-ci.org/nno/fieldtrip/builds/196540885 (scroll to the bottom to see a summary of the test results)
  * https://app.shippable.com/runs/588f1a6a6076a90f00241863/1/tests (click on 'tests' on the left hand side to see results from non-passing tests)
  These can be setup to run automatically for every push and/or pull request.
- In addition to passed or failed status, tests can be marked as skipped (based on walltime, memory, presence of dccnpath, or other dependencies) or errored.
- For failed and errored tests, the error message and stack trace are shown.
- There are some minor changes to existing code to suppress long outputs.
- The testing code is added in ``contrib/MOxUnit_fieldrip``. Tests for the testing code is in ``contrib/MOxUnit_fieldrip/tests`` and is tested through the CI services.

Currently no line coverage is implemented, although this is a feature that may be added in the future using MOcov. Using LIMITS="'maxwalltime',600,'maxmem','1gb'", currently the statistics are: 

- 42 passing
- 29 failing
- 1 error
- 522 skipped. 

Most tests are skipped due to memory limitations or the absence of 'dccnpath'.

robertoostenveld commented:
give me some time to look into this, I'll get back to you later today.

nno commented:
Link to bugzilla discussion: http://bugzilla.fieldtriptoolbox.org/show_bug.cgi?id=3066

nno commented:
Tests starting with "failed_" are now excluded by default when using ``ft_test moxunit_run``, but still marked as skipped.

nno commented:
@robertoostenveld thanks for your willingness to take a look. Please note the recent commits that better deal with excluding the "failed*.m" test files and not sending their results to the dashboard.

robertoostenveld commented:
I am having a look now. It is impressive! Nice to see how you have been able to pick up on the reimplemented fieldtrip dashboard (i.e. database backend and ft_test frontend).

I have to admit that I cannot follow everything, we might want to plan a Skype to talk things over. 

Let me comment some things here:

- moxunit_fieldtrip_util_parse_dependency needs some work. The idea is that lines starting with "% TEST" list the important functions on which it depends, either directly or indirectly. Developers know best what the critical functionality that is being tested. It would also be possible to use MATLAB depfun (now matlab.codetools.requiredFilesAndProducts), but I think that some non-machine intelligence works better. I have been thinking of renaming TEST to DEPENDENCY.

- we may want to avoid replication of json handling and of posting it. See compat/matlablt2015a/webwrite. 

- the test data files comprise 157G on our network file system at /home/common/matlab/fieldtrip/data/test. That is not all, since some tests use files from the tutorials and example scripts, which are at /home/common/matlab/fieldtrip/data/ftp. Together with some unzipped version of the tutorial data (which are in ./data itself)  the whole data is 224G. Sharing this is problematic due to its size, but also since not all data has been explicitly cleared for sharing/redistribution. Partial sharing might work, but synchronizing and versioning the test data is not simple.

robertoostenveld commented:
removing web('http://openmeeg.gforge.inria.fr') from om_checkombin is not OK. The website opening explicitly (and only once per session) is part of the agreement for including openmeeg in the FT release. I would be fine with adding a try-catch around it, so that it does not interfere with octave.

robertoostenveld commented:
I think it would be good to get the suggested changes to reduce the verbosity of existing FT code into the master asap. This includes mostly test scripts, getdimord and possibly some others. The reason for this is that I am presently also cleaning up test scripts independently, and would like to avoid conflicts and double work.

I don't know what the best approach is, I am still no git guru. What I would do is to make a list in a text file (using "git diff master --stat") of changed files. I would filter those for files that I am familiar with, I would merge those files from my "pull310" branch onto a clean branch, review that once more and then merge that into master. You could then merge with the master (or rebase on the master?) and continue working on the smaller set of test-specific functions. 

Does that sound like a plan? More sophisticated suggestions are welcome.

robertoostenveld commented:
Also something to discuss (probably best over Skype) is how to further develop 1) the MATLAB/Octave command line interface and 2) the web interface and wiki presence.

Re 1) My ideal is that if I am working on ft_whatever, I would do a commit in a branch, I would do
```
ft_test run dependency ft_whatever
```
and see what the existing tests do. With "ft_test compare" I would compare them to a previous run. After completing the work, but prior to merging a development branch into master (or a PR from someone else), I would want to do 
```
ft_test run
```
and again "ft_test compare" to check that no extra tests started failing.

Re 2) The dashboard web interface is now mostly absent. I would like to have some reflection of the testing process on http://www.fieldtriptoolbox.org/development/dashboard. I realize that - compared to subversion - the biggest challenge is that the git hashes are unpredictable. With svn I knew which revision we had, to which I could compare it (e.g .revison N could be compared to N-1 or N-5). But now it is random strings. A tree-like visualization of commit hashes and dates would already help. E.g. a clickable version of "git log --graph" as a wiki page. 

If I were to click on a hash on the graph page., I would like to go to the (tabular representation of) the known test results for that version. E.g. http://dashboard.fieldtriptoolbox.org/fieldtripversion/97dd149 would show the same as "ft_test report fieldtripversion 97dd149" in matlab. Advanced (development) interaction would remain on the command line (for quick interactive feedback), where the "compare" functionality is relevant.

robertoostenveld commented:
have a look at http://www.fieldtriptoolbox.org/development/dashboard for a status example of the graphical tree display. Rather than linking to github, it could link to dashboard.fieldtriptoolbox.org

nno commented:
Thanks @robertoostenveld for the detailed reply. It is probably a good idea to have a Skype session. Addressing your points:

> moxunit_fieldtrip_util_parse_dependency needs some work. The idea is that lines starting with "% TEST" list the important functions on which it depends, either directly or indirectly. Developers know best what the critical functionality that is being tested. It would also be possible to use MATLAB depfun (now matlab.codetools.requiredFilesAndProducts), but I think that some non-machine intelligence works better. I have been thinking of renaming TEST to DEPENDENCY.

What work would be needed on this function (except a possible change from TEST to DEPENDENCY in moxunit_fieldtrip_util_read_test_mfile_properties.m line 36)?

> we may want to avoid replication of json handling and of posting it. See compat/matlablt2015a/webwrite.

Good point, I will try to use that. 

> the test data files comprise 157G on our network file system at /home/common/matlab/fieldtrip/data/test. That is not all, since some tests use files from the tutorials and example scripts, which are at /home/common/matlab/fieldtrip/data/ftp. Together with some unzipped version of the tutorial data (which are in ./data itself) the whole data is 224G. Sharing this is problematic due to its size, but also since not all data has been explicitly cleared for sharing/redistribution. Partial sharing might work, but synchronizing and versioning the test data is not simple.

The large size and potential issues with distribution is quite unfortunate because it makes it difficult, if not impossible, to run the majority of the tests independently (outside of DCCN) . In addition the large size probably means also tests that take a long time to run, which is not great from a Continuous Integration perspective either. I understand that this is probably not a high priority, but I would definitely see the benefits of small data, or data that can be generated synthetically (for example see https://github.com/CoSMoMVPA/CoSMoMVPA/blob/master/mvpa/cosmo_synthetic_dataset.m). For now it would mean that tests with such data are skipped.
Are there plans to limit the size of test data and/or make such test data distributable for tests to be added in the future?

> removing web('http://openmeeg.gforge.inria.fr') from om_checkombin is not OK. The website opening explicitly (and only once per session) is part of the agreement for including openmeeg in the FT release. I would be fine with adding a try-catch around it, so that it does not interfere with octave.
> 
> also the pause in openmeeg_license should stay.

I moved the discussion for this issue to bugzilla: http://bugzilla.fieldtriptoolbox.org/show_bug.cgi?id=3240

> I think it would be good to get the suggested changes to reduce the verbosity of existing FT code into the master asap. This includes mostly test scripts, getdimord and possibly some others. The reason for this is that I am presently also cleaning up test scripts independently, and would like to avoid conflicts and double work.
> 
> I don't know what the best approach is, I am still no git guru. What I would do is to make a list in a text file (using "git diff master --stat") of changed files. I would filter those for files that I am familiar with, I would merge those files from my "pull310" branch onto a clean branch, review that once more and then merge that into master. You could then merge with the master (or rebase on the master?) and continue working on the smaller set of test-specific functions.
> 
> Does that sound like a plan? More sophisticated suggestions are welcome.

Your approach could work but it would remove the commit and author information, which is not nice to the original authors (me in this case). We can try ``git cherry-pick``. In the example below I took some of the commits that reduce the output from tests:
```
  git checkout master
  # make a new branch
  git checkout -b _msc/cherry_pick
  # use some other commits
  git cherry-pick 1db20263aa744e9c622ccfe2666372ec49e77681
  git cherry-pick cc8fc7b59c480fb1fd8f0edd13f532b8ca977708
  # we got a merge conflict... here, remove the offending file 
  git status
  git rm  test/test_ft_plot_mesh.m
  git cherry-pick --continue
  # use some other commits
  git cherry-pick 482d938878202f4fce45d215321c6d434d4127f6
  git cherry-pick be76890868a58d8f0bf9c12b3825292f8fa5ad35
  # now review the changes, run tests etc.
  # when happy then the branch can be pushed to GitHub for a PR and/or merged directly into master
```

> Also something to discuss (probably best over Skype) is how to further develop 1) the MATLAB/Octave command line interface and 2) the web interface and wiki presence.

Yes, this may be better to be discussed over Skype.

robertoostenveld commented:
I have stripped some of the comments below.
> 
> moxunit_fieldtrip_util_parse_dependency needs some work...
> 
> What work would be needed on this function (except a possible change from TEST to DEPENDENCY in moxunit_fieldtrip_util_read_test_mfile_properties.m line 36)?
> 

The detection of functions is not 100% correct, there are false alarms for variables that have the same name as functions, and primarily: indirectly called functions are not detected. Most of the high-level FT functions only do bookkeeping and outsource the actual work to low-level (private and module) functions. Changing a private function should be complemented by being able to run the tests that depend on that private function.


> ...
> 
> The large size and potential issues with distribution is quite unfortunate because it makes it difficult, if not impossible, to run the majority of the tests independently (outside of DCCN) . In addition the large size probably means also tests that take a long time to run, which is not great from a Continuous Integration perspective either.
> 
yes, some tests take close to 24 hours. I am not happy with that either. Not only the size of the datasets, but also the memory requirements of some tests is large and require serious compute nodes, which (I guess) don’t come for free with CI platforms. 

> I understand that this is probably not a high priority, but I would definitely see the benefits of small data, or data that can be generated synthetically (for example see https://github.com/CoSMoMVPA/CoSMoMVPA/blob/master/mvpa/cosmo_synthetic_dataset.m <https://github.com/CoSMoMVPA/CoSMoMVPA/blob/master/mvpa/cosmo_synthetic_dataset.m>). For now it would mean that tests with such data are skipped.
> Are there plans to limit the size of test data and/or make such test data distributable for tests to be added in the future?
> 
Plans yes, but priority not. It is more likely that we will simply start ignoring the long tests. We might look into some of them and break them up in shorter sections (to increase the turn-around time). Instructions to contributors are that - where possible- they should use synthetic data which is computed on the fly. But that does not always work. It is already hard as it is to get people to contribute more than mere statements like “it does not work, whereas it did last week”.

> …. Does that sound like a plan? More sophisticated suggestions are welcome.
> 
> Your approach could work but it would remove the commit and author information, which is not nice to the original authors (me in this case). We can try git cherry-pick. 
> 
Cool, I will try that! I was aware of the commit history getting lost and was hoping that you could teach me something new here.

nno commented:
>>> moxunit_fieldtrip_util_parse_dependency needs some work...

>> What work would be needed on this function (except a possible change from TEST to DEPENDENCY > in moxunit_fieldtrip_util_read_test_mfile_properties.m line 36)?

> The detection of functions is not 100% correct, there are false alarms for variables that have the same name as functions, and primarily: indirectly called functions are not detected. 

I tried to implement the same functionality as in ``ft_test_run``. Do you see differences between that function and what `moxunit_fieldtrip_util_read_test_mfile_properties.m` does? 

> Most of the high-level FT functions only do bookkeeping and outsource the actual work to low-level (private and module) functions. Changing a private function should be complemented by being able to run the tests that depend on that private function.

You would like functionality to be able to detect which functions are called, directly or indirectly, by a test function? It would make much more sense to me to actually have tests for the private functions, because it allows to actually test whether these functions still behave the same after changes.

If adding tests is not considered a priority, then for plan B: how would you detect which functions can be called directly or indirectly? It would seem it requires having a good Matlab parser to see which functions are called and would have to traverse over all files, and making a list recursively. It would be slow... not even talking about effects of path settings, or the current directory, or difficulties when using eval or functions called in mex files. Or am I missing something here?

robertoostenveld commented:
I have used SourceTree to cherry-pick the changes that only relate to the test scripts. 

See https://github.com/robertoostenveld/fieldtrip/tree/pull310-picked

I just started the test batch on our compute cluster. If that shows no errors, I will merge pull310-picked into master... some time has passed: 531 tests PASSED, 13 are still busy. That is good enough, I will merge it into master. Can you please rebase your branch against master?

I would also be fine to pick the changes that relate to the contrib/MOxUit_fieldtrip section. Easier however would be if that section were a separate pull request.

In general: I would like to see the changes that relate to the testing process separate from the changes that relate to test outcomes. Some changes that were made to the actual FT code might need some discussion, at least I want to be aware of the changes. Right now there are too many changes to get an easy overview.

robertoostenveld commented:
I just merged FT test-script specific changes into the master branch. 

There is also a change merged (to private plgndr.c) which is not consistent with that change.

````
mac011> git push upstream master
You're about to push master, is that what you intended? [y|n] y
Counting objects: 104, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (62/62), done.
Writing objects: 100% (104/104), 12.65 KiB | 0 bytes/s, done.
Total 104 (delta 86), reused 45 (delta 42)
remote: Resolving deltas: 100% (86/86), completed with 57 local objects.
To github.com:fieldtrip/fieldtrip.git
   1efe0e0..9696cf2  master -> master
````

@nno, can you rebase? I'll then have another look at the remaining changes.

nno commented:
@robertoostenveld thanks for the merging the cherry-picked commits. I have merged master and resolved the merge conflict. (Sorry for not rebasing - it may be a bit cleaner but usually just merge because I find that easier).

If you could have a look at the other commits that would be great. I would be happy to split up the commits in multiple separate PRs. Would you like to have a look first and maybe comment on what you think would be the best approach?

nno commented:
@robertoostenveld please note that I'm having some issues with compiling plgndr.c on the fly with Octave. That would definitely need some work before merging.

nno commented:
Travis update: 
FAILED (skipped=535, passed=51, failure=8)

https://travis-ci.org/nno/fieldtrip/jobs/198894027

robertoostenveld commented:
I have cherry-picked another 32 commits. 

````
mac011> git push upstream master
Counting objects: 183, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (122/122), done.
Writing objects: 100% (183/183), 50.41 KiB | 0 bytes/s, done.
Total 183 (delta 113), reused 96 (delta 61)
remote: Resolving deltas: 100% (113/113), completed with 3 local objects.
To github.com:fieldtrip/fieldtrip.git
   9696cf2..5520055  master -> master
````

It is getting more difficult now. There are some commits that have both test code and non-test code updates in them. I suppose that is the reason for .travis.yaml now being in conflict. 

My https://github.com/robertoostenveld/fieldtrip/tree/pull310 corresponds to your branch. My https://github.com/robertoostenveld/fieldtrip/tree/pull310-picked has been used to pick the selected commits. That one is now again merged in master.

There still appear to be 73 commits affecting 53 files in your branch that are different from the master. I don't think that a merge of master into your _wip/moxunit_fieldtrip_travis  will be enough. Please rebase your _wip/moxunit_fieldtrip_travis onto master. If that works, I can see whether I can get any further with the current procedure. If not, you will have to split it into separate sections.

If separate branches are needed, I would like to split it over
- test code
- test scripts
- mex and compilation related changes
- FT modules (fileio, specest, forward, plotting, etc)
- compat modules
- external (will require careful scrutiny)

nno commented:
@robertoostenveld thanks for merging the other commits. 
I have updated this branch to contain all commits from master up to and including https://github.com/fieldtrip/fieldtrip/commit/4c1c689d5a5e2b5a83401ac4bd584d2f3ee59414.

I would appreciate it you could do more cherry picking. Re the merge conflicts, I think have resolved all of them. Instead of using your pull310 branch, maybe it's better if you create a new branch based on the current  nno:_wip/moxunit_fieldtrip_travis branch?

If you could try to get any further that would be great. If not, please let me know and I will do my best to split up the commits in different PRs as you suggested.

robertoostenveld commented:
I did a visual inspection, it has cleaned up considerably and now "I can see the trees through the forest" ;-) 

I'll try to find time tomorrow to pull a fresh copy of your branch and continue merging.

nno commented:
@robertoostenveld any updates on this PR? It should be up-to-date with master.

robertoostenveld commented on ft_math.m:
```diff
@@ -480,6 +477,17 @@
 
+
+function assign_var(var, val)
+% Note: using an anonymous function as follows does not work in Octave:
```
can you document this in the code guidelines at http://www.fieldtriptoolbox.org/development/guidelines/code

robertoostenveld commented:
I reviewed it all: it looks fine and I'll merge. Tomorrow morning the test batch will have run through and we'll see if there is any damage (don't expect so).

nno commented:
@robertoostenveld thanks for the merge. I will write some documentation re Octave compatibility.

nno commented:
I have added the Octave-compatible alternative to 
```
    assign = @(var, val) assignin('caller', var, val);
```
to the guidelines at http://www.fieldtriptoolbox.org/development/guidelines/code
